flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/utils/TestingLogicalSlotProvider.java

```diff
@@ -53,30 +52,31 @@
 /**
  * A testing utility slot provider that simply has a predefined pool of slots.
  */
-public class SimpleSlotProvider implements SlotProvider, SlotOwner {
+public class TestingLogicalSlotProvider implements SlotProvider, SlotOwner {
 
 	private final Object lock = new Object();
 
 	private final ArrayDeque<SlotContext> slots;
 
 	private final HashMap<SlotRequestId, SlotContext> allocatedSlots;
 
-	public SimpleSlotProvider(JobID jobId, int numSlots) {
+	public TestingLogicalSlotProvider(JobID jobId, int numSlots) {
 		this(jobId, numSlots, new SimpleAckingTaskManagerGateway());
 	}
 
-	public SimpleSlotProvider(JobID jobId, int numSlots, TaskManagerGateway taskManagerGateway) {
+	public TestingLogicalSlotProvider(JobID jobId, int numSlots, TaskManagerGateway taskManagerGateway) {
 		checkNotNull(jobId, "jobId");
 		checkArgument(numSlots >= 0, "numSlots must be >= 0");
 
 		this.slots = new ArrayDeque<>(numSlots);
 
 		for (int i = 0; i < numSlots; i++) {
-			SimpleSlotContext as = new SimpleSlotContext(
+			SettableSlotContext as = new SettableSlotContext(
```

Comment:
Do we actually need to keep `SlotContext` in `slots`? Could it be just immediately `LogicalSlot`? That might simplify everything and then the `TestingLogicalSlot/SettableSlotContext` would not need to be created all the time.

Reply:
As for instancing `LogicalSlot` with the same physical info but not the same object, we need some thing like `SlotContext`. Please refer the discussion with @tillrohrmann above.

Reply:
ok, it seems we need to recreate at least `TestingLogicalSlot` to reset `releaseFuture`.